### PR TITLE
[cuda/cudnn] upgrade to 10.2

### PR DESCRIPTION
**Pull request by cenit**
As per title

Closes: #11341

### CODE REVIEW

**Comment by cenit** [comment]
@BillyONeal I took the opportunity also here to upgrade cuda on windows vm. Linux was already to cuda 10.2
There might also be a discrepancy somewhere, because the cmake script finds cuda 10.2 already also on windows...

File: ports/cudnn/portfile.cmake [hunk 1]
```diff
@@ -11,16 +11,16 @@ endif()
 vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY ONLY_DYNAMIC_CRT)
```
Comment (LilyWangL) [inline ports/cudnn/portfile.cmake:11]: `include(vcpkg_common_functions)` is deprecated. Please remove it. Please use command [vcpkg_fail_port_install](https://github.com/microsoft/vcpkg/blob/5deea3b975fe62990973d73a102de818d83ba20c/docs/maintainers/vcpkg_fail_port_install.md) to handle unsupported triplet.

File: ports/cudnn/CONTROL [hunk 2]
```diff
@@ -1,5 +1,5 @@
 Source: cudnn
-Version: 7.6
+Version: 7.6.5-CUDA10.2
 Description: NVIDIA's cuDNN deep neural network acceleration library
```
Comment (LilyWangL) [inline ports/cudnn/CONTROL:3]: Can you please add the Homepage for this port in the CONTROL file?

File: ports/cuda/portfile.cmake [hunk 3]
```diff
@@ -3,7 +3,12 @@
 # If this package is installed, we assume that CUDA is properly installed.
 
 #note: this port must be kept in sync with CUDNN port: every time one is upgraded, the other must be too
-set(CUDA_REQUIRED_VERSION "10.1.0")
+set(CUDA_REQUIRED_VERSION_MAJOR "10")
+set(CUDA_REQUIRED_VERSION_MINOR "2")
+
+
```
Comment (LilyWangL) [inline ports/cuda/portfile.cmake:9]: Please remove the extra blank lines.

**Review by LilyWangL — CHANGES_REQUESTED** [review]

File: ports/cudnn/portfile.cmake (hunk 1 above)
Comment (cenit) [reply ports/cudnn/portfile.cmake:11]: vcpkg_fail_port_install, introduces few months ago, is already being considered for deprecation. It’s introduction was not well considered and has many limitations. I will do your requests anyway

File: ports/cuda/portfile.cmake (hunk 3 above)
Comment (cenit) [reply ports/cuda/portfile.cmake:9]: I wanted to highlight lines to touch and lines to keep pristine. In a future update, we should touch only major and minor version and everything should be left as is.
I will remove spaces anyway, that was just the reason

File: ports/cuda/portfile.cmake [hunk 4]
```diff
@@ -3,7 +3,13 @@
 # If this package is installed, we assume that CUDA is properly installed.
 
 #note: this port must be kept in sync with CUDNN port: every time one is upgraded, the other must be too
-set(CUDA_REQUIRED_VERSION "10.1.0")
+set(CUDA_REQUIRED_VERSION_MAJOR "10")
+set(CUDA_REQUIRED_VERSION_MINOR "2")
+set(CUDA_REQUIRED_VERSION "${CUDA_REQUIRED_VERSION_MAJOR}.${CUDA_REQUIRED_VERSION_MINOR}.0")
```
Comment (simogasp) [inline ports/cuda/portfile.cmake:8]:
```suggestion
set(CUDA_REQUIRED_VERSION "10.2.0")
```

File: ports/cuda/portfile.cmake [hunk 5]
```diff
@@ -72,7 +78,7 @@ set(CUDA_VERSION_MAJOR ${CMAKE_MATCH_1})
 set(CUDA_VERSION_MINOR ${CMAKE_MATCH_2})
 set(CUDA_VERSION_PATCH ${CMAKE_MATCH_3})
 
-if (CUDA_VERSION_MAJOR LESS 10 AND CUDA_VERSION_MINOR LESS 1)
+if (CUDA_VERSION_MAJOR LESS CUDA_REQUIRED_VERSION_MAJOR AND CUDA_VERSION_MINOR LESS CUDA_REQUIRED_VERSION_MINOR)
```
Comment (simogasp) [inline ports/cuda/portfile.cmake:81]:
```suggestion
if (CUDA_VERSION VERSION_LESS ${CUDA_REQUIRED_VERSION})
```

File: ports/cuda/portfile.cmake [hunk 6]
```diff
@@ -72,7 +78,7 @@ set(CUDA_VERSION_MAJOR ${CMAKE_MATCH_1})
 set(CUDA_VERSION_MINOR ${CMAKE_MATCH_2})
```
Comment (simogasp) [inline ports/cuda/portfile.cmake:78]: i would change the 4 lines 75-78

```
string(REGEX MATCH "V([0-9]+)\\.([0-9]+)\\.([0-9]+)" CUDA_VERSION ${NVCC_OUTPUT})
message(STATUS "Found CUDA ${CUDA_VERSION}")
set(CUDA_VERSION_MAJOR ${CMAKE_MATCH_1})
set(CUDA_VERSION_MINOR ${CMAKE_MATCH_2})
```
into

```cmake
string(REGEX MATCH "V(([0-9]+)\\.([0-9]+)\\.([0-9]+))" CUDA_MATCH ${NVCC_OUTPUT})
set(CUDA_VERSION ${CMAKE_MATCH_1})
message(STATUS "Found CUDA ${CUDA_VERSION}")

**Review by simogasp — COMMENTED** [review]
Hi Stefano, here are the changes I made in my or #11341 

Basically I'd keep the variable `CUDA_REQUIRED_VERSION` and extract the full cuda version from the pattern VX.Y.Z output by nvcc.

Your solution goes in the same directions as mine (maintainability and avoid repeating hardcoded values), it just uses less code/variables 

If you are up to it, I will drop these changes in my PR.

File: ports/cuda/portfile.cmake (hunk 6 above)
Comment (cenit) [reply ports/cuda/portfile.cmake:76]: if you want to write it as a review i'd gladly accept it so that it belongs to you :)

**Comment by cenit** [comment]
Considering that i'd push for CUDA 11 and CUDNN 8 as soon as possible, I hope that it might be considered ready as soon as possible

**Comment by cenit** [comment]
@simogasp please let me know if you have any other suggestions

**Comment by simogasp** [comment]
it looks fine to me, thanks!

**Comment by cenit** [comment]
Do you mind closing your PR? Just to avoid confusion? Or do you think it’s better to keep both?

**Comment by simogasp** [comment]
I just closed because it turned out that the other thing I was fixed has been already merged in by another pr.

**Comment by BillyONeal** [comment]
The installer change fails to deploy:

```
Downloading CUDA...
Installing CUDA...
InstallCuda : Failed to install CUDA! Installation failed! Exited with -522190823.
At C:\Users\Administrator\Desktop\provision-image.ps1:449 char:1
+ InstallCuda -Url $CudaUrl -Features $CudaFeatures
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,InstallCuda
```

That error code is one neither I nor [Decoder💍](https://james.darpinian.com/decoder/?q=-522190823) recognize -- any ideas?

**Comment by BillyONeal** [comment]
I cleaned this PR into a single commit with most of the changes plus one that fixes the other HTTP links but don't want to force push into your branch; if you want that you can reset to this branch: https://github.com/BillyONeal/vcpkg/tree/cuda_for_cenit

**Comment by cenit** [comment]
Feel free to force push into my pr 

To be honest, I saw the same error on appveyor when deploying cuda10.2
Ended up installing cuda10.2 using chocolatey, which worked. I will inspect choco install script to see what they do

**Comment by cenit** [comment]
Error is due to failure when trying to install nvidia drivers. Which is expected, since there is no gpu in that machine set. This driver installation is unexpected and might be triggered by some cuda modules. On chocolatey it seems that they install cuda passing only the /s silent switch and it works. It also provides all the modules I’d expect. Shall we try in that way?

**Comment by BillyONeal** [comment]
@JackBoosY I don't believe anything is changed about this PR since I ran into the deploy failure.

**Comment by BillyONeal** [comment]
@cenit I was just about to kick off a new VM update and I see this is still unedited; are you still persuing it?

You can test changes yourself by getting a blank Server 2019 VM and running `$/scripts/azure-pipelines/windows/provision-image.txt` (renamed to .ps1) inside; we have fixed all the bits that required parameters to just not do those parts if not present.

**Comment by cenit** [comment]
Hi @BillyONeal 
Considering CUDA 11 is already released, I will prepare a PR for that as soon as possible. 
Considering cuda 10.2, which might be interesting to have in the meantime, If you mind waiting a little bit, I forgot to try installing cuda 10.2 in a disposable vm (as you say) “without specifying packages”. It looks like it might simply work (it’s a hint I received, untested)

**Review by strega-nil — APPROVED** [review]

**Comment by LilyWangL** [comment]
@cenit Can you please resolve the conflicts against master?

**Comment by BillyONeal** [comment]
@cenit Do you still want this open? It seems like any update in this area should go all the way to CUDA 11 update 1

**Pull request closed without merging**